contracts/vesting.token/src/vesting.token.cpp

```diff
@@ -0,0 +1,114 @@
+#include <vesting.token/vesting.token.hpp>
+
+namespace vestingtoken {
+
+    void vestingToken::updatedate(string sales_date_str, string launch_date_str) {
+        require_auth(get_self());
+        
+        vestingtoken::vesting_settings dates;
+        dates.sales_start_date = eosio::time_point_sec::from_iso_string(sales_date_str);
+        dates.launch_date = eosio::time_point_sec::from_iso_string(launch_date_str);
+
+        vestingtoken::settings launch_sales_dates_singleton(get_self(), get_self().value);
+        launch_sales_dates_singleton.set(dates, get_self());
+    }
+
+    void vestingToken::assigntokens(eosio::name holder, eosio::asset amount, int category_id) {
+        // Only the contract owner can call this function
+        require_auth(get_self());
+        
+        // Check if the provided category exists in the map
+        auto category_iter = vesting_categories.find(category_id);
+        eosio::check(category_iter != vesting_categories.end(), "Invalid vesting category");
+
+        // If the category exists, you can access it like this:
+        vesting_category category = category_iter->second;
+
+
+        // Check the symbol is correct and valid
+        auto sym = amount.symbol;
+        eosio::check(sym == system_resource_currency, "Symbol does not match system resource currency");
+        eosio::check(sym.precision() == system_resource_currency.precision(), "Symbol precision does not match");
+        eosio::check(amount.amount > 0, "Amount must be greater than 0");
+
+        // Create a new vesting schedule
+        vesting_allocations vesting_table(get_self(), holder.value);
+
+        // Check the number of rows for the user
+        uint32_t num_rows = 0;
+        for (auto itr = vesting_table.begin(); itr != vesting_table.end(); ++itr) {
+            ++num_rows;
+            if (num_rows > 20) {
+                eosio::check(false, "Cannot purchase tokens more than 20 times.");
+            }
+        }
+
+        // Calculate the number of seconds since sales start
+        uint32_t now_since_epoch = eosio::current_time_point().sec_since_epoch();
+        vestingtoken::settings launch_sales_dates_singleton(get_self(), get_self().value);
+        
+        vestingtoken::vesting_settings dates = launch_sales_dates_singleton.get();
+        uint32_t sales_start_date_since_epoch = dates.sales_start_date.sec_since_epoch();
+        
+        uint32_t allocated_after_sales_start_seconds = 0;
+        if (now_since_epoch > sales_start_date_since_epoch) {
+            allocated_after_sales_start_seconds = now_since_epoch - sales_start_date_since_epoch;
+        }
+
+        vesting_table.emplace(get_self(), [&](auto& row) {
+            row.holder = holder;
+            row.total_allocated = amount;
+            row.tokens_claimed = asset(0, amount.symbol);
+            row.seconds_since_sales_start = allocated_after_sales_start_seconds;
+            row.vesting_category_type = category;
+        });
+    }
```

Comment:
We need to ensure that the vesting contract has got the correct amount of LEOS issued to it (so that it can be withdrawn). Right now we don't have this logic here, so if we allocate tokens, there is no way to know if the contract holds enough funds for it to pay the withdrawals.

A few ways to do this
1. when we allocate the token, we need to transfer the LEOS to the contract (see below)
2. when we withdraw, we can issue the tokens (but then we need to ensure that we do not withdraw too many). i don't think this is a good idea

this needs to be easy enough for us to execute within the bootstrap script (so we can allocate all the LEOS to the correct vesting pools), and easy to then withdraw to the correct account that it belongs to


I think we also need to ensure we do a transfer here at the end of the action.
```cpp
eosio::action({get_self(), "active"_n},
            token_contract_name,
            "transfer"_n,
            std::make_tuple(sales_pool, get_self(), tokens_claimed , std::string("Unlocked vested coins")))
            .send();
        }
```

sales_pool should be an `eosio::name`, a variable in the vesting schedule

Reply:
is this tokens_claimed we need to transfer or token_allocated? sales_pool is which account?

Reply:
ok. so we agreed in call to have new argument `name sender` and at the end of the function we 
```
transfer:
- from: sender
- to: get_self()
- amount: amount
- memo: "Allocated vested funds"
- ```